x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/ChoroplethMap/index.tsx

```diff
@@ -0,0 +1,278 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, {
+  useState,
+  useEffect,
+  useRef,
+  useCallback,
+  useMemo
+} from 'react';
+import { Map, NavigationControl, Popup } from 'mapbox-gl';
+import { GeoJsonProperties } from 'geojson';
+import euiLightVars from '@elastic/eui/dist/eui_theme_light.json';
+import { shade, tint } from 'polished';
+import { Legend } from './Legend';
+import { ChoroplethToolTip } from './ChoroplethToolTip';
+
+interface ChoroplethDataElement {
+  key: string;
+  value: number;
+  docCount: any;
+}
+
+interface Tooltip {
+  name: string;
+  value: number;
+  docCount: number;
+}
+
+interface Props {
+  data: ChoroplethDataElement[];
+}
+
+const CHOROPLETH_LAYER_ID = 'choropleth_layer';
+const CHOROPLETH_POLYGONS_SOURCE_ID = 'choropleth_polygons';
+const GEOJSON_KEY_PROPERTY = 'iso2';
+const MAPBOX_STYLE =
+  'https://tiles.maps.elastic.co/styles/osm-bright-desaturated/style.json';
+const GEOJSON_SOURCE =
+  'https://vector.maps.elastic.co/files/world_countries_v1.geo.json?elastic_tile_service_tos=agree&my_app_name=ems-landing&my_app_version=7.2.0';
+
+export function getProgressionColor(scale: number) {
+  const baseColor = euiLightVars.euiColorPrimary;
+  const adjustedScale = 0.75 * scale + 0.05; // prevents pure black & white as min/max colors.
+  if (adjustedScale < 0.5) {
+    return tint(adjustedScale * 2, baseColor);
+  }
+  if (adjustedScale > 0.5) {
+    return shade(1 - (adjustedScale - 0.5) * 2, baseColor);
+  }
+  return baseColor;
+}
+
+const getMin = (items: ChoroplethDataElement[]) =>
+  Math.min(...items.map(item => item.value));
+
+const getMax = (items: ChoroplethDataElement[]) =>
+  Math.max(...items.map(item => item.value));
+
+export const ChoroplethMap: React.SFC<Props> = props => {
+  const { data } = props;
+
+  const containerRef = useRef<HTMLDivElement>(null);
+  const [map, setMap] = useState<Map | null>(null);
+  const popupRef = useRef<Popup | null>(null);
+  const popupContainerRef = useRef<HTMLDivElement>(null);
+  const enableScrollZoom = useRef(false);
+  const [tooltipState, setTooltipState] = useState<Tooltip | null>(null);
+  const [min, max] = useMemo(() => {
+    const minValue = getMin(data);
+    const maxValue = getMax(data);
+    return [
+      Number.isFinite(minValue) ? minValue : 0,
+      Number.isFinite(maxValue) ? maxValue : 0
```

Comment:
Isn't this only relevant when `data`/`items` is empty? And in those cases do we even want to show the legend?

Reply:
Ah, good point. Didn't realized that. I think that's okay for now - the legend currently looks like this without any data so I think it's better to not show it:

<img width="1106" alt="Screen Shot 2019-08-29 at 09 18 08" src="https://user-images.githubusercontent.com/209966/63918939-04debc80-ca3e-11e9-9c84-f48b9b3026b4.png">